### DISCUSSION TITLE
How to get translation string from *.dll

### DISCUSSION THREAD

**Discussion by steam3d** [microsoft/microsoft-ui-xaml#8960] [Q&A] [answered]
I created Class Library (WinUl 3 in Desktop) and some UI components inside.  I can easely use `x:Uid` to translate them from XAML, but from code when I could not get access to this strings. 

I tried this, but It trying to get string from my application instead of dll resources

```
 var resourceLoader = Windows.ApplicationModel.Resources.ResourceLoader.GetForViewIndependentUse();
 var name = resourceLoader.GetString("String_Name");
```

`String_Name` exists only in dll en-US Resources.resw.  When I run code above a get nothing. 

So how to get translation string inside dll from code?

**Comment by DarranRowe** [comment]
First of all, desktop applications, are recommended to use MRT Core, not MRT. This means that you want to use [Microsoft.Windows.ApplicationModel.Resources.ResourceLoader](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.applicationmodel.resources.resourceloader?view=windows-app-sdk-1.4) and related functionality.
Secondly, are you sure that the resources do not contain that string? You can check the individual .resw files to see if the string has been placed in these files. If a resource with the name String_Name is present in each of these files, you should then check the .pri file to see if String_Name has been correctly merged into the .pri file. The .pri file that you want to check depends, if your class library is in its own .msix package, you will want to check the resources.pri file of that package. If your class library is part of an application, then you will want to check either resources.pri of the application package if it is a packaged application, or [executable base name].pri if it is unpackaged. As an example of what I mean, if the application's executable name is myapp.exe, then you will want to check myapp.pri if it is an unpackaged application.
You will need the Windows SDK tools available if you want to check the .pri file. The makepri tool can be used to dump the contents of the file. So you can use the command `makepri dump /if <path to .pri file> /of <path to output with a .xml extension>`. So `makepri dump /if myapp.pri /of myapp.pri.xml` is an example if you run it from the same directory as myapp.pri.
To show a practical example of this. Suppose that I am using some resources:

![Screenshot 2023-10-10 133556](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/40d1f4c5-d3a0-4580-8c5c-c5951e0775f6)

![Screenshot 2023-10-10 133607](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/ff7288cb-aaa7-44de-a76e-3538350c1903)

![Screenshot 2023-10-10 133633](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/c720fbb3-146a-4614-bab0-117470e576cd)

![Screenshot 2023-10-10 133642](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/a644d9ba-bd39-4862-9c99-8110d5c588f9)

There are two text resources here. One is in the executable's resources itself where the other is in a component's resources. Since this application is unpackaged, I am looking at Meh.pri here, since the application's executable is Meh.exe.

`C:\Users\Darran\source\repos\Meh\x64\Debug\Meh>makepri dump /if meh.pri /of meh.pri.xml`

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PriInfo>
	<ResourceMap name="Meh" version="1.0" primary="true">
		<Qualifiers>
			<Language>EN-GB,JA-JP</Language>
		</Qualifiers>
		<ResourceMapSubtree name="Files">
			<NamedResource name="App.xbf" uri="ms-resource://Meh/Files/App.xbf">
				<Candidate type="Path">
					<Value>App.xbf</Value>
				</Candidate>
			</NamedResource>
			<NamedResource name="MainWindow.xbf" uri="ms-resource://Meh/Files/MainWindow.xbf">
				<Candidate type="Path">
					<Value>MainWindow.xbf</Value>
				</Candidate>
			</NamedResource>
			<ResourceMapSubtree name="TestComponent">
				<NamedResource name="TestPage.xbf" uri="ms-resource://Meh/Files/TestComponent/TestPage.xbf">
					<Candidate type="Path">
						<Value>TestComponent\TestPage.xbf</Value>
					</Candidate>
				</NamedResource>
			</ResourceMapSubtree>
		</ResourceMapSubtree>
		<ResourceMapSubtree name="Strings">
			<NamedResource name="Greeting" uri="ms-resource://Meh/Strings/Greeting">
				<Candidate qualifiers="Language-JA-JP" type="String">
					<Value>こんにちは</Value>
				</Candidate>
				<Candidate qualifiers="Language-EN-GB" isDefault="true" type="String">
					<Value>Hello</Value>
				</Candidate>
			</NamedResource>
		</ResourceMapSubtree>
		<ResourceMapSubtree name="TestComponent">
			<ResourceMapSubtree name="Strings">
				<NamedResource name="Test" uri="ms-resource://Meh/TestComponent/Strings/Test">
					<Candidate qualifiers="Language-JA-JP" type="String">
						<Value>テスト</Value>
					</Candidate>
					<Candidate qualifiers="Language-EN-GB" isDefault="true" type="String">
						<Value>Test</Value>
					</Candidate>
				</NamedResource>
			</ResourceMapSubtree>
		</ResourceMapSubtree>
	</ResourceMap>
</PriInfo>
```

Right, so how do you obtain this? Well, the dumped .pri above shows the full path to the resource. It is also based upon the path of the resource. In the earlier screenshots, do you notice how Strings.resw is in a Strings/[language] subdirectory of the project? That is going to form part of the path. On top of that, if it is the resources for a component then the component name will be added to the path.

```
    MainWindow::MainWindow()
    {
        InitializeComponent();

        auto mgr = winrt::Microsoft::Windows::ApplicationModel::Resources::ResourceManager();
        auto rm = mgr.MainResourceMap();
        auto v = rm.TryGetValue(LR"(TestComponent\Strings\Test)");
        if (v)
        {
            Texty().Text(v.ValueAsString());
        }
    }
```

![Screenshot 2023-10-10 135603](https://github.com/microsoft/microsoft-ui-xaml/assets/52577874/1b1b669f-8d5a-4def-bd82-aa7294e4c261)

The above code, while C++, works properly. It should also be easy to translate this to C#.

**Reply by steam3d** [reply]
In my case *.pri files contains:

resources.pri

```
<NamedResource name="String_Name" uri="ms-resource://9b9c3c99-8a69-4fcd-9ce9-2883b7f3a869/SDK/Resources/String_Name">
...
</NamedResource> 
```

sdk.pri

```
<NamedResource name="String_Name" uri="ms-resource://SDK/SDK/Resources/String_Name">
...
</NamedResource>
```
So instead of `SDK\Strings\String_Name` I have to use `SDK\Resources\String_Name`.   I'm not sure, if I write this everywhere in the dll library code, it will work in all projects at least because you use `String`, and I use `Resources`

If I have both translation from main project and sdk dll marged together in resources.pri why this code crush with exception `System.Runtime.InteropServices.COMException: 'NamedResource not found.`

```
ResourceLoader resourceLoader = new ResourceLoader();
var name = resourceLoader.GetString("String_Name");
```

It's all so confusing. Why do I need to use relative paths when I'm trying to get a dll resource from a dll code.

**Reply by steam3d** [reply]
Thanks for the help. Your answer prompted the correct query. I figured out the problem, but I'm still puzzled as to why Microsoft decided to go this route.

**Comment by steam3d** [comment, marked as answer]
**WindowsAppSDK-Samples** contains necessary info. Check: [ResourceManagement](https://github.com/microsoft/WindowsAppSDK-Samples/tree/main/Samples/ResourceManagement)
